File: sdk/csharp/src/Uba.Core/Batcher.cs
using System;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;

namespace Uba
{
    /// <summary>flush 结果</summary>
    public class FlushResult
    {
        public bool Success { get; set; }
        public PostReportResponse? Response { get; set; }
        /// <summary>丢弃的事件数（重试耗尽）</summary>
        public int Dropped { get; set; }
    }

    /// <summary>
    /// 事件缓冲与批量合并：内存队列，达到 batchSize 或 flushInterval 触发上报。
    /// 上报失败重试耗尽时丢弃事件（防内存无限增长）。
    /// </summary>
    public class Batcher
    {
        private readonly UbaConfig _cfg;
        private readonly IHttpTransport _transport;
        private readonly Func<ClientInfo?> _getClientInfo;
        private readonly Action<LogLevel, string> _log;

        private readonly List<ReportEvent> _queue = new List<ReportEvent>();
        private readonly object _lock = new object();
        private Timer? _timer;
        private int _flushing; // 0/1 标志，防并发 flush

        public Batcher(UbaConfig cfg, IHttpTransport transport, Func<ClientInfo?> getClientInfo, Action<LogLevel, string> log)
        {
            _cfg = cfg;
            _transport = transport;
            _getClientInfo = getClientInfo;
            _log = log;
            StartTimer();
        }

        /// <summary>入队一条事件</summary>
        public void Enqueue(ReportEvent evt)
        {
            lock (_lock)
            {
                _queue.Add(evt);
                if (_queue.Count >= _cfg.BatchSize)
                {
                    // 不阻塞调用方，触发异步 flush
                    _ = FlushAsync();
                }
            }
        }

        public int Size
        {
            get { lock (_lock) return _queue.Count; }
        }

        /// <summary>触发批量上报</summary>
        public async Task<FlushResult> FlushAsync()
        {
            // 防并发：已有 flush 在进行则直接返回
            if (Interlocked.CompareExchange(ref _flushing, 1, 0) != 0)
                return new FlushResult { Success = true };

            List<ReportEvent> events;
            lock (_lock)
            {
                if (_queue.Count == 0)
                {
                    _flushing = 0;
                    return new FlushResult { Success = true };
                }
                events = new List<ReportEvent>(_queue);
                _queue.Clear();
            }

            try
            {
                return await SendWithRetryAsync(events).ConfigureAwait(false);
            }
            finally
            {
                _flushing = 0;
            }
        }

        private async Task<FlushResult> SendWithRetryAsync(List<ReportEvent> events)
        {
            var req = new PostReportRequest
            {
                AppId = _cfg.AppId,
                AppSecret = _cfg.AppSecret,
                Events = events,
                ClientInfo = _getClientInfo(),
            };
            string body = JsonSerializer.Serialize(req);
            string url = _cfg.Endpoint + _cfg.Path;

            FetchResult last;
            for (int attempt = 0; attempt <= _cfg.MaxRetries; attempt++)
            {
                last = await _transport.SendAsync(url, body, _cfg.Timeout).ConfigureAwait(false);

                if (last.Ok)
                {
                    if (last.Response != null && (last.Response.FailedCount ?? 0) > 0)
                    {
                        _log(LogLevel.Warn, $"upload partial failure: success={last.Response.SuccessCount} failed={last.Response.FailedCount}");
                    }
                    return new FlushResult { Success = true, Response = last.Response };
                }

                // 4xx（非 429）客户端错误：不重试
                if (last.Status >= 400 && last.Status < 500 && last.Status != 429)
                {
                    _log(LogLevel.Error, $"upload rejected (no retry): {Summarize(last)}");
                    return new FlushResult { Success = false, Dropped = events.Count };
                }

                if (attempt < _cfg.MaxRetries)
                {
                    int delay = _cfg.RetryBaseDelay * (int)Math.Pow(2, attempt);
                    _log(LogLevel.Warn, $"upload failed (attempt {attempt + 1}), retry in {delay}ms: {Summarize(last)}");
                    await Task.Delay(delay).ConfigureAwait(false);
                }
                else
                {
                    _log(LogLevel.Error, $"dropping {events.Count} events after { _cfg.MaxRetries + 1} attempts: {Summarize(last)}");
                    return new FlushResult { Success = false, Dropped = events.Count };
                }
            }
            return new FlushResult { Success = false, Dropped = events.Count };
        }

        private static string Summarize(FetchResult r)
        {
            if (r.Exception != null) return r.Exception;
            if (r.Error != null) return $"status={r.Status} reason={r.Error.Reason} msg={r.Error.Message}";
            return $"status={r.Status}";
        }

        private void StartTimer()
        {
            _timer = new Timer(_ => { _ = FlushAsync(); }, null, _cfg.FlushInterval, _cfg.FlushInterval);
        }

        /// <summary>销毁：停止定时器</summary>
        public void Dispose()
        {
            _timer?.Dispose();
            _timer = null;
        }
    }
}

File: frontend/sdk/web/uba/src/batcher.ts
/**
 * 事件缓冲与批量合并
 *
 * 内存队列缓冲事件，达到 batchSize 或 flushInterval 触发批量上报。
 * 上报失败且重试耗尽时，事件被丢弃（避免内存无限增长）。
 */

import { sendWithRetry, sendBeacon, type FetchResult } from './retry';
import type { ReportEvent, PostReportRequest, PostReportResponse, ClientInfo } from './types';

export interface BatcherOptions {
  appId: string;
  appSecret: string;
  /** 完整上报 URL */
  url: string;
  /** 批量大小阈值 */
  batchSize: number;
  /** 定时 flush 间隔（毫秒） */
  flushInterval: number;
  /** 透传给 sendWithRetry 的配置 */
  maxRetries: number;
  timeout: number;
  baseDelay: number;
  /** 是否启用 beacon 兜底 */
  enableBeacon: boolean;
  /** clientInfo 注入器（每次 flush 取最新值） */
  getClientInfo: () => ClientInfo | undefined;
  /** 日志 */
  log: (level: 'warn' | 'error' | 'info', msg: string) => void;
}

export interface FlushResult {
  success: boolean;
  response?: PostReportResponse;
  /** 丢弃的事件数（重试耗尽后） */
  dropped: number;
}

export class Batcher {
  private queue: ReportEvent[] = [];
  private timer: ReturnType<typeof setTimeout> | null = null;
  private flushing = false;
  /** 上报中文档化请求构造所需的常量 */

  constructor(private opts: BatcherOptions) {
    this.startTimer();
  }

  /** 入队一条事件，满 batchSize 自动 flush */
  enqueue(event: ReportEvent): void {
    this.queue.push(event);
    if (this.queue.length >= this.opts.batchSize) {
      void this.flush();
    }
  }

  /** 当前队列长度 */
  size(): number {
    return this.queue.length;
  }

  /**
   * 触发批量上报。从队列取出当前所有事件发送。
   * 多次并发调用只会执行一次实际发送（其余等待）。
   */
  async flush(): Promise<FlushResult> {
    if (this.flushing) {
      return { success: true, dropped: 0 };
    }
    if (this.queue.length === 0) {
      return { success: true, dropped: 0 };
    }

    this.flushing = true;
    const events = this.queue.splice(0, this.queue.length);

    try {
      const body = this.buildBody(events);
      const result = await sendWithRetry(this.opts.url, body, {
        maxRetries: this.opts.maxRetries,
        timeout: this.opts.timeout,
        baseDelay: this.opts.baseDelay,
        log: (level, msg) => this.opts.log(level, msg),
      });

      if (result.ok) {
        // 注意：HTTP 200 也可能含部分失败事件（errorsByType），这里记录但不视为整体失败
        const resp = result.response;
        if (resp && resp.failedCount && resp.failedCount > 0) {
          this.opts.log(
            'warn',
            `upload partial failure: success=${resp.successCount} failed=${resp.failedCount}`,
          );
        }
        return { success: true, response: resp, dropped: 0 };
      }

      // 发送失败（含重试耗尽）：事件被丢弃
      this.opts.log('error', `dropping ${events.length} events due to upload failure`);
      return { success: false, dropped: events.length };
    } finally {
      this.flushing = false;
    }
  }

  /**
   * 页面卸载兜底：用 sendBeacon 尽力投递当前队列。
   * 不等待响应，best-effort。
   */
  flushBeacon(): void {
    if (!this.opts.enableBeacon || this.queue.length === 0) {
      return;
    }
    const events = this.queue.splice(0, this.queue.length);
    const body = this.buildBody(events);
    const ok = sendBeacon(this.opts.url, body);
    if (!ok) {
      this.opts.log('warn', 'sendBeacon failed, events lost on unload');
    }
  }

  /** 销毁：停止定时器 */
  destroy(): void {
    this.stopTimer();
  }

  private buildBody(events: ReportEvent[]): string {
    const req: PostReportRequest = {
      appId: this.opts.appId,
      appSecret: this.opts.appSecret,
      events,
      clientInfo: this.opts.getClientInfo(),
    };
    return JSON.stringify(req);
  }

  private startTimer(): void {
    this.stopTimer();
    this.timer = setInterval(() => {
      void this.flush();
    }, this.opts.flushInterval);
    // Node 下 setInterval 返回 NodeJS.Timeout，浏览器返回 number；不阻塞进程退出
    if (this.timer && typeof (this.timer as any).unref === 'function') {
      (this.timer as any).unref();
    }
  }

  private stopTimer(): void {
    if (this.timer) {
      clearInterval(this.timer);
      this.timer = null;
    }
  }
}

/** 重新导出，便于上层使用 */
export type { FetchResult };
